test/e2e/registration_service_test.go

```diff
@@ -429,6 +431,177 @@ func (s *registrationServiceTestSuite) TestSignupOK() {
+
+	mp := make(map[string]interface{})
+	err = json.Unmarshal([]byte(body), &mp)
+	require.NoError(s.T(), err)
+
+	mpStatus, ok := mp["status"].(map[string]interface{})
+	assert.True(s.T(), ok)
+
+	assert.Equal(s.T(), "", mp["compliantUsername"])
+	assert.Equal(s.T(), identity0.Username, mp["username"])
+	require.IsType(s.T(), false, mpStatus["ready"])
+	assert.False(s.T(), mpStatus["ready"].(bool))
+	assert.Equal(s.T(), "PendingApproval", mpStatus["reason"])
+	require.True(s.T(), mpStatus["verificationRequired"].(bool))
+
+	// Approve usersignup.
+	userSignup.Spec.Approved = true
+	err = s.hostAwait.Client.Update(context.TODO(), userSignup)
+	require.NoError(s.T(), err)
+
+	// Ensure the Master User Record is NOT provisioned
+	_, err = s.hostAwait.WaitForMasterUserRecord(identity0.Username)
+	require.Error(s.T(), err)
+
+	// Initiate the verification process
+	requestBody := `{ "country_code":"+61", "phone_number":"408999999" }`
+	req, err = http.NewRequest("PUT", s.route+"/api/v1/signup/verification",
+		strings.NewReader(requestBody))
+	require.NoError(s.T(), err)
+	req.Header.Set("Authorization", "Bearer "+token0)
+	req.Header.Set("content-type", "application/json")
+	resp, err = httpClient.Do(req)
+	require.NoError(s.T(), err)
+	defer close(s.T(), resp)
+
+	body, err = ioutil.ReadAll(resp.Body)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), body)
+	assert.Equal(s.T(), http.StatusNoContent, resp.StatusCode)
+
+	// Retrieve the updated UserSignup
+	userSignup, err = s.hostAwait.WaitForUserSignup(identity0.ID.String())
+	require.NoError(s.T(), err)
+
+	// Confirm there is a verification code label
+	require.NotEmpty(s.T(), userSignup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey])
+
+	// Confirm the expiry time has been set
+	require.NotEmpty(s.T(), userSignup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+
+	// Attempt to verify with an incorrect verification code
+	req, err = http.NewRequest("GET", s.route+"/api/v1/signup/verification/invalid",
+		strings.NewReader(requestBody))
+	require.NoError(s.T(), err)
+	req.Header.Set("Authorization", "Bearer "+token0)
+	req.Header.Set("content-type", "application/json")
+	resp, err = httpClient.Do(req)
+	require.NoError(s.T(), err)
+	defer close(s.T(), resp)
+
+	body, err = ioutil.ReadAll(resp.Body)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), body)
+	require.Equal(s.T(), http.StatusForbidden, resp.StatusCode)
+
+	// Retrieve the updated UserSignup
+	userSignup, err = s.hostAwait.WaitForUserSignup(identity0.ID.String())
+	require.NoError(s.T(), err)
+
+	// Check attempts has been incremented
+	require.NotEmpty(s.T(), userSignup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+
+	// Verify with the correct code
+	// Attempt to verify with an incorrect verification code
+	req, err = http.NewRequest("GET", s.route + fmt.Sprintf("/api/v1/signup/verification/%s",
+		userSignup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey]),
```

Comment:
Let's also verify that the initial verification code from line 520 is still the same in UserSignup. Let's keep the code from line 520 and check that the same code in the update usersignup from line 541.
It would test that we do not update the code if verification failed.

Reply:
Done.